### ISSUE TITLE
some C/C++ warnings not working correctly when type/declaration and usage/implementation are in different files

### ISSUE THREAD

**Issue by AlexRouSg** [clangd/clangd#1429]
There are some C/C++ warnings/diagnostics that would require processing more than the currently opened file to be useful. 

Examples:

For `-Wpadded` just having the following in the header will not trigger any warnings:
```
struct padded{
	int i;
	bool b;
	float f;
};
```
However it will work if we use it in the same file like: 
`constexpr auto was_padded = sizeof(padded) != sizeof(int) + sizeof(bool) + sizeof(float);`
If the type and usage are in separate files then there is no warning.

---

Another example would be `-Wunused-function`/`-Wunused-variable`
If we have a `static inline void foo{}` or a `static int bar = 0` in a header file but usage is elsewhere, it will always warn that it is unused.

---

**System information**

Output of `clangd --version`:
`clangd version 15.0.6 (https://github.com/llvm/llvm-project 088f33605d8a61ff519c580a71b1dd57d16a03f8)`
Editor/LSP plugin:
`vscode + llvm-vs-code-extensions.vscode-clangd`
Operating system:
`windows 10/ubuntu 22.04`

**Comment by HighCommander4** [comment]
The `-Wpadded` diagnostic is dropped due to [this check](https://searchfox.org/llvm/rev/1f057e365f1fdd630c023a990e84e95a6c792e4d/clang-tools-extra/clangd/Diagnostics.cpp#879). It arises while processing the file containing the `sizeof(padded)`, which is the "main file" in this code, but neither the diagnostic nor any attached notes are in the main file.

**Comment by HighCommander4** [comment]
Also relevant to mention is [this check](https://searchfox.org/llvm/rev/1f057e365f1fdd630c023a990e84e95a6c792e4d/clang-tools-extra/clangd/Diagnostics.cpp#870), which tries to transform a diagnostic located in an included file to a diagnostic located in the main file at the point of the #include, with a note in the original location in the header -- but it only does this for errors, not warnings.

**Comment by HighCommander4** [comment]
So I believe the idea here is:

 * For a warning in an included file (unlike for an error), we don't want to show it in the main file at the point of the #include because that would be quite noisy.
 * But we also don't bother to issue it at the original location because (1) if the header is not open in the editor, that wouldn't do much (it might show up in a Problems view if the editor has one, but it's unclear that polluting those with warnings from unopened headers included by open files is desired), and (2) if the header is open in the editor, we assume that the diagnostic would be produced when building the header's own AST.

It's this latter assumption that breaks down with `-Wpadded` -- it's produced during the build of the AST of the including file that's triggering layout of the structure with `sizeof(padded)`, but it has no location tying it to that trigger point.

It's not clear to me how to fix this without regressing other aspects of behaviour described above. If we had some way to tell that the `-Wpadded` diagnostic was "produced by" code in the main file, we could condition on that and emit the diagnostic in its original location (i.e. in the header), but I don't see such a way.

**Comment by HighCommander4** [comment]
I guess one resolution approach could be: if a diagnostic is produced by the main file AST build (not the preamble build), and it doesn't mention the main file, emit it at its original location anyways. I think that shouldn't produce too much noise because most warnings in included header files should be produced by the preamble build.

**Comment by sam-mccall** [comment]
> if a diagnostic is produced by the main file AST build (not the preamble build), and it doesn't mention the main file, emit it at its original location anyways.

LSP requires us to emit all the diagnostics for a file at once, they replace the previous set of diagnostics for that file.
Currently we satisfy this by providing diagnostics for the main (open) file only.

Emitting diagnostics for headers in the preamble seems possible, but it's a large/significant change (likely performance implications on both client and server, what should the default thresholds be, what to do if different main files produce different diagnostics for headers, etc)

**Comment by palapapa** [comment]
I don't know if this is related to my issue. I have a `app.hpp` that's included in `main.cpp`. In `app.hpp` there is this line in a class: `static constexpr auto REQUIRED_DEVICE_FEATURES_12 = VkPhysicalDeviceVulkan12Features { .timelineSemaphore = VK_TRUE, .scalarBlockLayout = VK_TRUE, .bufferDeviceAddress = VK_TRUE };`. This gives the `ISO C++ requires field designators to be specified in declaration order; field 'timelineSemaphore' will be initialized after field 'scalarBlockLayout'` warning when compiled, but clangd doesn't report it. The clangd logs say `V[00:03:52.538] Dropped diagnostic: : ISO C++ requires field designators to be specified in declaration order; field 'timelineSemaphore' will be initialized after field 'scalarBlockLayout'`.

**Comment by AlexRouSg** [comment]
@palapapa your issue is that if a struct is defined as:
```
struct {
int A;
int B;
int C;
}
```
when initialising you must do:
```
auto foo = {.A = 0, .B = 1, .C = 2};
```

what you did was:
```
auto foo = {.B = 1, .C = 2, .A = 0, };
```
which is not allowed

**Comment by palapapa** [comment]
I know, the problem is that clangd doesn't report it, which sounds similar to your issue.

**Comment by HighCommander4** [comment]
Yes, it sounds similar (though in this case it's less obvious to me why the diagnostic can't be produced while processing the header code only).
